packages/remix-react/browser.tsx

```diff
@@ -48,6 +48,19 @@ declare global {
 
 let router: Router;
 let hmrAbortController: AbortController | undefined;
+let hmrRouterReadyResolve: ((router: Router) => void) | undefined;
+// There's a race condition with HMR where the remix:manifest is signaled before
+// the router is assigned in the RemixBrowser component. This promise gates the
+// HMR handler until the router is ready
+let hmrRouterReadyPromise = new Promise<Router>((resolve) => {
+  // body of a promise is executed immediately, so this can be resolved outside
+  // of the promise body
+  hmrRouterReadyResolve = resolve;
+}).catch(() => {
+  // This is a noop catch handler to avoid unhandled promise rejection warnings
+  // in the console. The promise is never rejected.
+  return undefined;
+});
```

Comment:
You could use a `Channel` here instead so you don't have to manually set `resolve` and check for it later to make TS happy:

```suggestion
let routerReady = Channel.create<Router>()
// ...
let checkRouter = await routerReady.result()
if (!checkRouter.ok) {
  console.error("...")
  return
}
let router = checkRouter.value
// ...

routerReady.ok(router)
```

Reply:
That's fine with me, however, it looks like that Channel implementation is in remix-dev package?
https://github.com/remix-run/remix/blob/0d22ee92450eb6caf78055d3d9d956b244de6519/packages/remix-dev/channel.ts#L28

And it doesn't look like there's an import in this react package, so not sure how y'all would want to do that.

Also, typescript seems happy with it for me! I'm getting all the proper types out of this.

Reply:
Oh yea good point. Not worth introducing the channel abstraction just for this if its not already in this package.

Reply:
Ahh, got it! 
Yeah, I really like that Channel abstraction, probably going to shamelessly use it somewhere else 😂